File: python/condition_group_tokenizer.py
from enum import Enum

class EnumConditionGroupToken(Enum):
    '''分词Token枚举'''
    Undefined = "忽略空格、换行符"
    BracketLeft = "左括号"
    BracketRight = "右括号"
    BracketContent = "括号内容"
    Or = "或运算"
    And = "且运算"
    Content = "具体条件内容"

class ConditionGroupToken():
    '''分词Token类'''
    def __init__(self, token:EnumConditionGroupToken, source:str, depth:int) -> None:
        self.token = token
        self.source = source
        self.depth = depth

class ConditionGroupTokenizer():
    '''条件组分词器'''
    def isIgnoreChar(self, ch:str) -> bool:
        if ch == ' ' or ch == '\n':
            return True
        return False

    def proxyCondition(self, source:str, isPromt:bool) -> bool:
        """
        测试例子，可以重载该方法
        """
        temp = source[0:-1]
        if temp == "true":
            return True
        else:
            return False

    def isConditionCmd(self, cmd:ConditionGroupToken, depth:int, isPromt:bool) -> bool:
        if cmd.token == EnumConditionGroupToken.Content:
            return self.proxyCondition(cmd.source, isPromt)
        elif cmd.token == EnumConditionGroupToken.BracketContent:
            return self.directCheckConditionGroup(cmd.source, depth + 1, isPromt)

    def directCheckConditionGroup(self, source:str, depth:int = 0, isPromt:bool = False) -> bool:
        result:list[ConditionGroupToken] = []
        oldToken = EnumConditionGroupToken.Undefined
        newToken = oldToken

        length = len(source)
        bracketMathTimes = 0
        temp = ""
        for i in range(length):
            ch = source[i]

            # 忽略字符
            if self.isIgnoreChar(ch) and bracketMathTimes == 0:
                newToken = EnumConditionGroupToken.Undefined
            elif ch == '(':
                if bracketMathTimes == 0:
                    newToken = EnumConditionGroupToken.BracketLeft
                else:
                    newToken = EnumConditionGroupToken.BracketContent
                bracketMathTimes += 1
            elif ch == ')' and bracketMathTimes > 0:
                bracketMathTimes -= 1
                if bracketMathTimes == 0:
                    newToken = EnumConditionGroupToken.BracketRight
                else:
                    newToken = EnumConditionGroupToken.BracketContent
            elif ch == '&' and bracketMathTimes == 0:
                newToken = EnumConditionGroupToken.And
            elif ch == '|' and bracketMathTimes == 0:
                newToken = EnumConditionGroupToken.Or
            elif bracketMathTimes > 0:
                newToken = EnumConditionGroupToken.BracketContent
            else:
                newToken = EnumConditionGroupToken.Content

            if newToken != oldToken:
                if len(temp) > 0:
                    if oldToken != EnumConditionGroupToken.BracketLeft and oldToken != EnumConditionGroupToken.BracketRight:
                        result.append(ConditionGroupToken(oldToken, temp, depth))
                    temp = ""
                oldToken = newToken

            if newToken != EnumConditionGroupToken.Undefined:
                temp += ch

        if len(temp) > 0:
            if oldToken == EnumConditionGroupToken.Content:
                result.append(ConditionGroupToken(oldToken, temp, depth))

        # 检测条件是否满足
        finalResult = self.isConditionCmd(result[0], depth, isPromt)

        for i in range(len(result) - 1):
            cmd = result[i]
            if cmd.token == EnumConditionGroupToken.And:
                finalResult = finalResult and self.isConditionCmd(result[i + 1], depth, isPromt)
            elif cmd.token == EnumConditionGroupToken.Or:
                finalResult = finalResult or self.isConditionCmd(result[i + 1], depth, isPromt)

        return finalResult

    def directCheck(self, source:str, isPromt:bool = False) -> bool:
        return self.directCheckConditionGroup(source, 0, isPromt)

File: cplusplus/ConditionGroupTokenizer.cpp
#include "ConditionGroupTokenizer.h"

bool ConditionGroupTokenizer::IsIgnoreChar(char ch) {
    return ch == ' ' || ch == '\n';
}

void ConditionGroupTokenizer::ParseConditionGroup(const std::string & source, std::vector<ConditionBlock> &blockList, int depth)
{
    EnumConditionGroupToken oldToken = EnumConditionGroupToken::None;
    EnumConditionGroupToken newToken = oldToken;
    unsigned int length = source.length();
    std::string temp;
    int bracketMatchTimes = 0;
    for (int i = 0; i < length; i++)
    {
        char ch = source[i];

        // 忽略字符
        if (IsIgnoreChar(ch) && bracketMatchTimes == 0)
        {
            newToken = EnumConditionGroupToken::None;
        }
        else if (ch == '(')
        {
            if (bracketMatchTimes == 0)
            {
                newToken = EnumConditionGroupToken::BracketLeft;
            }
            else
            {
                newToken = EnumConditionGroupToken::Content;
            }
            bracketMatchTimes++;
        }
        else if (ch == ')' && bracketMatchTimes > 0)
        {
            bracketMatchTimes--;
            if (bracketMatchTimes == 0)
            {
                newToken = EnumConditionGroupToken::BracketRight;
            }
            else
            {
                newToken = EnumConditionGroupToken::Content;
            }
        }
        else if (ch == '&' && bracketMatchTimes == 0)
        {
            newToken = EnumConditionGroupToken::And;
        }
        else if (ch == '|' && bracketMatchTimes == 0)
        {
            newToken = EnumConditionGroupToken::Or;
        }
        else
        {
            newToken = EnumConditionGroupToken::Content;
        }

        if (oldToken != newToken)
        {
            if (temp.length() > 0)
            {
                // 小括号模块则递归判断
                if (oldToken == EnumConditionGroupToken::Content && newToken == EnumConditionGroupToken::BracketRight)
                {
                    ParseConditionGroup(temp, blockList, depth + 1);
                }
                else {
                    ConditionBlock obj;
                    obj.token = oldToken;
                    obj.data = temp;
                    obj.depth = depth;
                    blockList.push_back(obj);
                }

                temp = "";
            }

            oldToken = newToken;
        }

        if (newToken != EnumConditionGroupToken::None)
            temp += ch;
    }

    if (temp.length() > 0)
    {
        if (oldToken == EnumConditionGroupToken::Content)
        {
            ConditionBlock obj;
            obj.token = oldToken;
            obj.data = temp;
            obj.depth = depth;
            blockList.push_back(obj);
        }
    }
}

bool ConditionGroupTokenizer::ProxyCondition(const std::string &source, bool isPrompt) {
        if (isPrompt)
        {
            printf("Start Logic Check: %s\n", source.c_str());
        }

        std::string str = source.substr(0, source.length() - 1);
        if (str == "false")
            return false;
        else
            return  true;
    }

    bool ConditionGroupTokenizer::RunConditionExpression(const std::vector<ConditionBlock> &blockList, int depth, bool isPrompt) {
    bool finalResult = false;
    EnumConditionGroupToken currentSymbol = EnumConditionGroupToken::None;
    for (int i = 0; i < blockList.size(); i++) {
        const ConditionBlock &block = blockList[i];
        if (block.depth == depth) {
            switch (block.token) {
                case EnumConditionGroupToken::And:
                case EnumConditionGroupToken::Or:
                    currentSymbol = block.token;
                    break;
                case EnumConditionGroupToken::Content: {
                    if (currentSymbol != EnumConditionGroupToken::None) {
                        switch (currentSymbol) {
                            case EnumConditionGroupToken::And:
                                finalResult = finalResult && ProxyCondition(block.data, isPrompt);
                                break;
                            case EnumConditionGroupToken::Or:
                                finalResult = finalResult || ProxyCondition(block.data, isPrompt);
                                break;
                            default:
                                break;
                        }
                    } else {
                        finalResult = ProxyCondition(block.data, isPrompt);
                    }
                }
                    break;
                case EnumConditionGroupToken::BracketLeft: {
                    int bracketL = i;
                    int bracketR = bracketL;
                    for (; bracketR < blockList.size(); bracketR++) {
                        const ConditionBlock &temp = blockList[bracketR];
                        if (temp.token == EnumConditionGroupToken::BracketRight && temp.depth == depth)
                            break;
                    }

                    std::vector<ConditionBlock> bracketList(blockList.begin() + (bracketL + 1),
                                                            blockList.begin() + bracketR);
                    if (currentSymbol != EnumConditionGroupToken::None)
                    {
                        switch (currentSymbol) {
                            case EnumConditionGroupToken::And:
                                finalResult = finalResult && RunConditionExpression(bracketList, depth + 1, isPrompt);
                                break;
                            case EnumConditionGroupToken::Or:
                                finalResult = finalResult || RunConditionExpression(bracketList, depth + 1, isPrompt);
                                break;
                            default:
                                break;
                        }
                    }else{
                        finalResult = RunConditionExpression(bracketList, depth + 1, isPrompt);
                    }
                    i = bracketR;
                }
                    break;
                default:
                    break;
            }
        }
    }

    return finalResult;
}

bool ConditionGroupTokenizer::RunConditionExpression(bool isPrompt) {
    if (!conditionBlocks.empty())
    {
        return RunConditionExpression(conditionBlocks, 0, isPrompt);
    }

    return true;
}

bool ConditionGroupTokenizer::Check(const std::string& source) {
    conditionBlocks.clear();
    conditionHashSet.clear();
    ParseConditionGroup(source, conditionBlocks);
    for (std::vector<ConditionBlock>::iterator pos = conditionBlocks.begin(); pos != conditionBlocks.end(); ++pos)
    {
        if (pos->token == EnumConditionGroupToken::Content)
            conditionHashSet.insert(pos->data);
    }

    return RunConditionExpression(true);
}

bool ConditionGroupTokenizer::DirectCheck(const std::string &source) {
    return DirectCheckConditionGroup(source);
}

bool ConditionGroupTokenizer::DirectCheckConditionGroup(const std::string &source, int depth)
{
    std::vector<ConditionBlock> result;
    EnumConditionGroupToken oldToken = EnumConditionGroupToken::None;
    EnumConditionGroupToken newToken = oldToken;
    unsigned int lenght = source.length();
    std::string temp;
    int bracketMatchTimes = 0;
    for (int i = 0; i < lenght; i++)
    {
        char ch = source[i];

        // 忽略字符
        if (IsIgnoreChar(ch) && bracketMatchTimes == 0)
        {
            newToken = EnumConditionGroupToken::None;
        }
        else if (ch == '(')
        {
            if (bracketMatchTimes == 0)
            {
                newToken = EnumConditionGroupToken::BracketLeft;
            }
            else
            {
                newToken = EnumConditionGroupToken::BracketContent;
            }
            bracketMatchTimes++;
        }
        else if (ch == ')' && bracketMatchTimes > 0)
        {
            bracketMatchTimes--;
            if (bracketMatchTimes == 0)
            {
                newToken = EnumConditionGroupToken::BracketRight;
            }
            else
            {
                newToken = EnumConditionGroupToken::BracketContent;
            }
        }
        else if (ch == '&' && bracketMatchTimes == 0)
        {
            newToken = EnumConditionGroupToken::And;
        }
        else if (ch == '|' && bracketMatchTimes == 0)
        {
            newToken = EnumConditionGroupToken::Or;
        }
        else if (bracketMatchTimes > 0)
        {
            newToken = EnumConditionGroupToken::BracketContent;
        }
        else
        {
            newToken = EnumConditionGroupToken::Content;
        }

        if (oldToken != newToken)
        {
            if (temp.length() > 0)
            {
                if (oldToken != EnumConditionGroupToken::BracketLeft && oldToken != EnumConditionGroupToken::BracketRight)
                {
                    ConditionBlock obj;
                    obj.token = oldToken;
                    obj.data = temp;
                    obj.depth = depth;
                    result.push_back(obj);
                }
                temp = "";
            }

            oldToken = newToken;
        }

        if (newToken != EnumConditionGroupToken::None)
            temp += ch;
    }

    if (temp.length() > 0)
    {
        if (oldToken == EnumConditionGroupToken::Content)
        {
            ConditionBlock obj;
            obj.token = oldToken;
            obj.data = temp;
            obj.depth = depth;
            result.push_back(obj);
        }
    }

    bool finalResult = IsConditionCmd(result[0], depth);
    for (int i = 0; i < result.size(); i++)
    {
        ConditionBlock &cmd = result[i];
        switch (cmd.token)
        {
            case EnumConditionGroupToken::And:
                finalResult = finalResult && IsConditionCmd(result[i + 1], depth);
                break;
            case EnumConditionGroupToken::Or:
                    finalResult = finalResult || IsConditionCmd(result[i + 1], depth);
                    break;
            default:
                break;
        }
    }
    return finalResult;
}

bool ConditionGroupTokenizer::IsConditionCmd(const ConditionBlock &cmd, int depth) {
    switch (cmd.token)
    {
        case EnumConditionGroupToken::Content:
            return ProxyCondition(cmd.data, true);
        case EnumConditionGroupToken::BracketContent:
                return DirectCheckConditionGroup(cmd.data, depth + 1);
        default:
            break;
    }

    return false;
}
